### PR TITLE
BUG: Pickle NA objects

### LINKED ISSUE #31847: BUG: pd.NA doesn't pickle/unpickle faithfully
#### Code Sample, a copy-pastable example if possible

```python

In [5]: df['Gold Categories'].count()
Out[5]: 135218

In [6]: df['Gold Categories'].isna().sum()
Out[6]: 0

In [7]: df['Gold Categories'].iloc[256]
Out[7]: <NA>

In [8]: pd.isna(df['Gold Categories'].iloc[256])
Out[8]: False

In [9]: type(df['Gold Categories'].iloc[256])
Out[9]: pandas._libs.missing.NAType

In [10]: pd.__version__
Out[10]: '1.0.1'


```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.5.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.3.16-200.fc30.x86_64
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : nb_NO.UTF-8
LOCALE           : nb_NO.UTF-8

pandas           : 1.0.1
numpy            : 1.17.3
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 19.3.1
setuptools       : 41.6.0.post20191030
Cython           : 0.29.13
pytest           : 5.2.2
hypothesis       : None
sphinx           : 2.2.1
blosc            : None
feather          : None
xlsxwriter       : 1.2.2
lxml.etree       : 4.4.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : 2.8.4 (dt dec pq3 ext lo64)
jinja2           : 2.10.3
IPython          : 7.9.0
pandas_datareader: None
bs4              : 4.8.1
bottleneck       : 1.2.1
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.4.1
matplotlib       : 2.2.3
numexpr          : 2.7.0
odfpy            : None
openpyxl         : 3.0.0
pandas_gbq       : None
pyarrow          : 0.15.1
pytables         : None
pytest           : 5.2.2
pyxlsb           : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.10
tables           : 3.5.2
tabulate         : 0.8.5
xarray           : None
xlrd             : 1.2.0
xlwt             : 1.3.0
xlsxwriter       : 1.2.2
numba            : 0.46.0

</details>

### PR DESCRIPTION
**Pull request by TomAugspurger**
According to
https://docs.python.org/3/library/pickle.html#object.__reduce__,

> If a string is returned, the string should be interpreted as the name
> of a global variable. It should be the object’s local name relative to
> its module; the pickle module searches the module namespace to determine
> the object’s module. This behaviour is typically useful for singletons.

Closes https://github.com/pandas-dev/pandas/issues/31847

### CODE REVIEW

File: pandas/tests/scalar/test_na_scalar.py [hunk 1]
```diff
@@ -267,3 +269,9 @@ def test_integer_hash_collision_set():
     assert len(result) == 2
     assert NA in result
     assert hash(NA) in result
+
+
+def test_pickle_roundtrip():
```
Comment (jreback) [inline pandas/tests/scalar/test_na_scalar.py]: also can u test with read/to_pickle

File: pandas/tests/scalar/test_na_scalar.py [hunk 2]
```diff
@@ -267,3 +269,29 @@ def test_integer_hash_collision_set():
...
+    # https://github.com/pandas-dev/pandas/issues/31847
+    result = pickle.loads(pickle.dumps(pd.NA))
+    assert result is pd.NA
+
+
+def test_pickle_roundtrip_pandas():
+    with tm.ensure_clean("data.pkl") as f:
```
Comment (jreback) [inline pandas/tests/scalar/test_na_scalar.py]: there are test functions to do this in pandas._testing

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jorisvandenbossche — APPROVED** [review]
Thanks!

File: pandas/tests/scalar/test_na_scalar.py [hunk 3]
```diff
@@ -267,3 +269,29 @@ def test_integer_hash_collision_set():
...
+        with open(f):
+            pd.to_pickle(pd.NA, f)
+        with open(f):
+            result = pd.read_pickle(f)
+    assert result is pd.NA
+
+
+def test_pickle_roundtrip_series():
+    s = pd.Series(pd.array([1, 2, pd.NA]))
```
Comment (jorisvandenbossche) [inline pandas/tests/scalar/test_na_scalar.py]: Can you also add a test with object array + one with string array? As the integer array doesn't actually store pd.NA, while those others do

Comment (jorisvandenbossche) [reply pandas/tests/scalar/test_na_scalar.py]: (to also cover https://github.com/pandas-dev/pandas/issues/32095, and https://github.com/pandas-dev/pandas/issues/31847#issuecomment-588500110)

**Comment by jbrockmendel** [comment]
@TomAugspurger can you re-push?  looks like azure has stalled

**Comment by TomAugspurger** [comment]
It'll be a week or so before I can look at this again.

File: pandas/tests/scalar/test_na_scalar.py [hunk 4]
```diff
@@ -267,3 +269,29 @@ def test_integer_hash_collision_set():
...
     assert hash(NA) in result
+
+
+def test_pickle_roundtrip():
+    # https://github.com/pandas-dev/pandas/issues/31847
+    result = pickle.loads(pickle.dumps(pd.NA))
+    assert result is pd.NA
+
+
+def test_pickle_roundtrip_pandas():
```
Comment (jreback) [inline pandas/tests/scalar/test_na_scalar.py]: ideally better to just parameterize this (@jorisvandenbossche comment below)

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@TomAugspurger can you rebase and update (small requests)

**Comment by TomAugspurger** [comment]
I think everything has been addressed.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks @TomAugspurger !

**Comment by jorisvandenbossche** [comment]
@MeeseeksDev backport to 1.0.x

**Pull request merged**